server/backend/api/alembic/versions/V6_add_role_type.py

```diff
@@ -0,0 +1,74 @@
+"""add_role_type
+
+Revision ID: V6
+Revises: V5
+Create Date: 2022-09-28 09:53:42.810469
+
+"""
+from alembic import op
+import sqlalchemy as sa
+import sqlmodel
+import app.models.model
+import datetime
+
+
+from sqlalchemy.dialects import postgresql
+
+# revision identifiers, used by Alembic.
+revision = 'V6'
+down_revision = 'V5'
+branch_labels = None
+depends_on = None
+
+# capture if offline as dates need to be treated differently for inserts if
+# offline
+from alembic import context
+migration_context = context.get_context()
+isOffline = migration_context.environment_context.is_offline_mode()
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('fam_role_type',
+    sa.Column('role_type_code', sa.String(length=2), nullable=False, comment='role type code'),
+    sa.Column('description', sa.String(length=100), nullable=True, comment='Description of what the role_type_code represents'),
+    sa.Column('effective_date', postgresql.TIMESTAMP(precision=6), nullable=False, comment='The date and time the code was effective.'),
+    sa.Column('expiry_date', postgresql.TIMESTAMP(precision=6), nullable=True, comment='The date and time the code expired.'),
+    sa.Column('update_date', postgresql.TIMESTAMP(precision=6), nullable=True, comment='The date and time the record was created or last updated.'),
+    sa.PrimaryKeyConstraint('role_type_code', name='fam_role_type_code_pk'),
+    schema='app_fam',
+    comment='A role type is a code that is associated with roles that will influence what can be associate with a role.  At time of implementation an abstract role can only have other roles related to it, while a concrete role can only have users associated with it'
+    )
+    op.add_column('fam_role', sa.Column('role_type_code', sa.String(length=2), nullable=False, comment='Identifies if the role is an abstract or concrete role. Users should only be assigned to roles where role_type=concrete'), schema='app_fam')
+    op.create_foreign_key('reffam_role_type', 'fam_role', 'fam_role_type', ['role_type_code'], ['role_type_code'], source_schema='app_fam', referent_schema='app_fam')
+    # ### end Alembic commands ###
+
+    famRoleType = app.models.model.FamRoleType.__table__
+    currentDate = datetime.datetime.now()
+    dateEntry = currentDate
+
+    if isOffline:
+        currentDateStr = currentDate.strftime("%Y-%m-%d %H:%M:%S")
+        dateEntry = op.inline_literal(currentDateStr)
+
```

Comment:
Hi Kevin,
Maybe I don't understand why 'isOffline' and online needs different 'dateEntry' value.
Would you explain a little? Thanks, just want to make sure I understand it.

Reply:
Hey @ianliuwk1019, so when we are doing development with the containerized postgres database, we can run the alembic migrations directly against the database using a command like: `alembic upgrade head` to run all migratiions, or for a specific version `alembic upgrade V5`.  

Then before we push the code up to github, and create a PR, we need to generate a flyway migration.  The script server/backend/api/alembic/generateFlywayMigrations.py basically runs a "offline" migration, which instead of executing the migration against a database, it outputs a SQL file that it would have normally sent to the database.  That SQL file becomes the flyway migration.

For some reason when the migration is run "offline" (with the --sql switch) I was getting errors when it tried to generate the insert statements that insert dates.  To resolve that, is why i needed to detect online vs offline migration.  

Hope that helps